### PR TITLE
Fix: Fixed an issue where it wasn't possible to edit some file permissions

### LINKED ISSUE #11549: Bug: Cannot change ACE type in SecurityAdvanced page
### Description

- Related: #11471 
- There's no functionality for this. No code.

### Steps To Reproduce

1. Go to file properties
2. Change tab to Security
3. Click '>' button to open SecurityAdvanced page
4. Change editable ACE type
5. Save
6. And you can see that it was not saved.

### Steps to fix

- Add functionality for them.

### Files Version

Version: 2.4.40 OS Version: 25281.1000

### Windows Version

Windows 11 Pro Insider Preview 22H2

### Log file

_None_

### LINKED ISSUE #11550: Bug: Cannot change access controls in SecurityAdvanced page
### Description

- Related: #11471 
- There's no functionality for this. No code.

### Steps To Reproduce

1. Go to file properties
2. Change tab to Security
3. Click '>' button to open SecurityAdvanced page
4. Change editable ACE's access controls
5. Save
6. And you can see that it was not saved.

### Steps to fix

- Add functionality for this.

### Files Version

Version: 2.4.40 OS Version: 25281.1000

### Windows Version

Windows 11 Pro Insider Preview 22H2

### Log file

_None_

### LINKED ISSUE #11551: Bug: Cannot change ACE's inheritance in SecurityAdvanced page
### Description

- Related: #11471 
- There's no functionality for this. No code.

### Steps To Reproduce

1. Go to file properties
2. Change tab to Security
3. Click '>' button to open SecurityAdvanced page
4. Change editable ACE's inheritance ('apply to')
5. Save
6. And you can see that it was not saved.

### Steps to fix

- Add functionality for this.

### Files Version

Version: 2.4.40 OS Version: 25281.1000

### Windows Version

Windows 11 Pro Insider Preview 22H2

### Log file

_None_

### LINKED ISSUE #12255: Bug: Cannot add/remove access control entry
### Description

As the title.

### Steps To Reproduce

1. Click a file's properties
2. Go to Security
3. Click + or -

or

1. Click a file's properties
2. Go to Security
3. Go to Advanced permissions
4. Click Change above.

### Requirements

_Non_

### Files Version

2.4.70.0

### Windows Version

10.0.25336.1010

### Log file

_None_

### PR DESCRIPTION
**Pull request by 0x5bfa**
### Description

Fix an issue which has been reported.

### Motivation and Context

I've been aware of that there're a lot of lack of function. As a part of improving, I fix an issue where some security functionalities are missing.

### Validation

- If you have a file with no ACEs
  - Remove all ACEs
  - Open its properties and see Security again
  - You can see a message that is telling there's no ACE
- If you have a file whose owner is not current user
  - Change owner to whatever that is not you like SYSTEM, Administrators
  - Open its properties and see Security again
  - You can see a message that is telling you do not have Read Permission access rights and to see ACEs, change owner.

### Removal & Its Rationale

- Removal of functionalities
  - I removed a functionality of changing access control type
  - I removed a functionality of changing access control flags
  - I removed a functionality of changing inheritance flags
  - I removed a functionality of converting inherited ACE to explicit one
  - I removed a functionality of removing inherited ACE
- Rationale of that removal
  - I will add back them in a further PR. Those functionalities are not at work long time before and so I decided those are might as well not to be.
- Removal of this message
- Rationale
  - File Explorer can see ACEs with admin privileges even if the current user doesn't have Read Permission. However, WinUI3 app cannot elevate current instance without no restarting, so there's no meaning that Continue button exists. In this case, in Files, Files suggest that you should take ownership of the object even if the current user can see ACEs with admin privileges like File Explorer do when even admin privilege doesn't work (for example, the owner is SYSTEM or something else that doesn't belong to Admin group).
  image

### PR Checklist
- [x] Were these changes approved in an issue or discussion with the project maintainers? In order to prevent extra work, feature requests and changes to the codebase must be approved before the pull request will be reviewed. This prevents extra work for the contributors and maintainers.
  - Related #11551 
  - Related #11550 
  - Related #11549
  - Related #11930 
  - Closes #12255
  - Related #11471
- [x] Did you build the app and test your changes?
- [x] Did you check for accessibility? You can use Accessibility Insights for this.
- [ ] Did you remove any strings from the en-us resource file?
   - [ ] Did you search the solution to see if the string is still being used? 
- [ ] Did you implement any design changes to an existing feature?
   - [ ] Was this change approved?
- [x] Are there any other steps that were used to validate these changes?
  - See Validation section.

### Screenshots

Before (File Explorer or previous Files)|After
---|---
image|image
image|image
image|image

### CODE REVIEW

**Comment by 0x5bfa** [comment]
Ready for review.

**Comment by 0x5bfa** [comment]
@yair100 I finished working on Security functionalities. can you review?

**Comment by 0x5bfa** [comment]
Done.

**Comment by 0x5bfa** [comment]
@QuaintMako Could you review?

**Comment by QuaintMako** [comment]
> @QuaintMako Could you review?

I should be able in the upcoming days, probably around Sunday at worst.

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs [hunk 1]
```diff
@@ -13,56 +13,144 @@
...
+			=> IsSelected && !IsInherited && false;
+
+		/// <summary>
+		/// AccessControlTypeHumanized
+		/// </summary>
+		public string AccessControlTypeHumanized
+			=> AccessControlType switch
+			{
+				AccessControlEntryType.Allow => "Allow",
+				_ => "Deny" // AccessControlType.Deny
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/AccessControlEntry.cs:49]: If we comment the value on the side, better put it in the switch directly?

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs [hunk 2]
```diff
@@ -13,56 +13,144 @@
...
+			};
+
+		/// <summary>
+		/// AccessControlTypeGlyph
+		/// </summary>
+		public string AccessControlTypeGlyph
+			=> AccessControlType switch
+			{
+				AccessControlEntryType.Allow => "\xE73E",
+				_ => "\xF140" // AccessControlType.Deny
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/AccessControlEntry.cs:59]: If we comment the value on the side, better put it in the switch directly?

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs [hunk 3]
```diff
@@ -318,40 +341,37 @@ public bool DeniedFullControlAccess
...
+			//});
 
-				InheritanceFlags = Enum.Parse<InheritanceFlags>(parts[0]);
-				PropagationFlags = Enum.Parse<PropagationFlags>(parts[1]);
-			});
+			//ChangeInheritanceFlagsCommand = new RelayCommand<string>(x =>
+			//{
+			//	var parts = x.Split(',');
+			//	InheritanceFlags = Enum.Parse<AccessControlEntryFlags>(parts[0]);
+			//});
```
Comment (QuaintMako) [inline src/Files.App/Filesystem/Security/AccessControlEntry.cs:357]: Remove commented code

File: src/Files.App/Helpers/FileSecurityHelpers.cs [hunk 4]
```diff
@@ -20,29 +14,11 @@ namespace Files.App.Helpers
...
-			}
-
-			return true;
-		}
-
+		/// <summary>
+		/// Get file owner.
+		/// </summary>
+		/// <param name="path"></param>
+		/// <returns></returns>
```
Comment (QuaintMako) [inline src/Files.App/Helpers/FileSecurityHelpers.cs:21]: Is this doc really necessary? It doesn't give much information that one can't get from the function name.

File: src/Files.App/Helpers/FileSecurityHelpers.cs [hunk 5]
```diff
@@ -60,166 +36,227 @@ public static string GetOwner(string path)
 			return szSid;
 		}
 
-		public static bool GetAccessControlProtection(string path, bool isFolder)
+		/// <summary>
+		/// Set file owner
+		/// </summary>
+		/// <param name="path"></param>
+		/// <param name="sid"></param>
+		/// <returns></returns>
```
Comment (QuaintMako) [inline src/Files.App/Helpers/FileSecurityHelpers.cs:44]: Same

File: src/Files.App/Helpers/FileSecurityHelpers.cs [hunk 6]
```diff
@@ -60,166 +36,227 @@ public static string GetOwner(string path)
...
+			var result = GetNamedSecurityInfo(
+				szPath,
 				SE_OBJECT_TYPE.SE_FILE_OBJECT,
 				SECURITY_INFORMATION.DACL_SECURITY_INFORMATION | SECURITY_INFORMATION.PROTECTED_DACL_SECURITY_INFORMATION,
 				out _,
 				out _,
-				out var pDacl,
+				out PACL pDACL,
 				out _,
 				out _);
```
Comment (QuaintMako) [inline src/Files.App/Helpers/FileSecurityHelpers.cs:182]: Could be one lined.

File: src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs [hunk 7]
```diff
@@ -1,218 +1,251 @@
...
-					RemoveAccessControlEntryCommand.NotifyCanExecuteChanged();
-					DisableInheritanceCommand.NotifyCanExecuteChanged();
-					ReplaceChildPermissionsCommand.NotifyCanExecuteChanged();
-				}
+				//if (!_isProtected)
+				//	return "SecurityAdvancedInheritedEnable/Text".GetLocalizedResource();
+				//else if (_preserveInheritance)
+				//	return "SecurityAdvancedInheritedConvert/Text".GetLocalizedResource();
+				//else
+				//	return "SecurityAdvancedInheritedRemove/Text".GetLocalizedResource();
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs:46]: Commented code, is there a reason to keep it? If so, add an explanation please.

File: src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs [hunk 8]
```diff
@@ -1,218 +1,251 @@
...
-			DisableInheritanceCommand = new RelayCommand(DisableInheritance, () => AccessControlList is not null && AccessControlList.IsValid && (AccessControlList.IsProtected != _isProtected));
-			SetDisableInheritanceOptionCommand = new RelayCommand<string>(SetDisableInheritanceOption);
-			ReplaceChildPermissionsCommand = new RelayCommand(ReplaceChildPermissions, () => AccessControlList is not null && AccessControlList.IsValid);
+			string imageres = System.IO.Path.Combine(CommonPaths.SystemRootPath, "System32", "imageres.dll");
+			var imageResList = Shell.Win32API.ExtractSelectedIconsFromDLL(
+				imageres,
+				new List<int>() { Constants.ImageRes.ShieldIcon },
+				16);
+
+			ShieldIconFileInfo = imageResList.First();
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs:164]: I did some work around this in #12068 that might be able to be in common for the shield icon.

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs [hunk 9]
```diff
@@ -32,88 +45,108 @@ public AccessControlList AccessControlList
...
+
+			var error = FileSecurityHelpers.GetAccessControlList(_path, _isFolder, out _AccessControlList);
+			_SelectedAccessControlEntry = AccessControlList.AccessControlEntries.FirstOrDefault();
+
+			if (!AccessControlList.IsValid)
+			{
+				DisplayElements = false;
+				ErrorMessage = error == Win32Error.ERROR_ACCESS_DENIED
+					? "You must have Read permissions to view the properties of this object. Click 'Advanced permissions' to continue."
+					: "Unable to display permissions for one or more errors";
```
Comment (QuaintMako) [inline src/Files.App/ViewModels/Properties/SecurityViewModel.cs:101]: Should be localized.

File: src/Files.App/Views/Properties/SecurityAdvancedPage.xaml [hunk 10]
```diff
@@ -639,12 +687,12 @@
...
 					x:Name="ReplaceChildPermissionsButton"
 					Grid.Column="1"
-					x:Load="{x:Bind SecurityAdvancedViewModel.IsFolder, Mode=OneWay}"
+					x:Load="{x:Bind SecurityAdvancedViewModel.AccessControlList.IsFolder, Mode=OneWay}"
 					Command="{x:Bind SecurityAdvancedViewModel.ReplaceChildPermissionsCommand, Mode=OneWay}"
 					ToolTipService.ToolTip="{helpers:ResourceString Name=SecurityAdvancedReplaceChildPermissions2/ToolTipService/ToolTip}"
 					Visibility="Collapsed">
 					<TextBlock Text="{helpers:ResourceString Name=SecurityAdvancedReplaceChildPermissions/Text}" TextTrimming="CharacterEllipsis" />
-				</Button>
+				</Button>-->
```
Comment (QuaintMako) [inline src/Files.App/Views/Properties/SecurityAdvancedPage.xaml:695]: A reason to keep the code commented from l.655 to l.695?

**Comment by QuaintMako** [comment]
Also, I **did not test any fonctionnality**, I'm not versed at all in the security processes.

File: src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs (hunk 8 above)
Comment (0x5bfa) [reply src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs:164]: I have to get it as byte[]. But that public helper method will return other than byte[].

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs [hunk 11]
```diff
@@ -13,56 +13,144 @@
...
+		/// <summary>
+		/// Whether the ACE is inherited or not
+		/// </summary>
+		public bool IsInherited { get; private set; }
+
+		/// <summary>
+		/// Whether the ACE is editable or not
+		/// </summary>
+		public bool IsEditable
+			=> IsSelected && !IsInherited && false;
```
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App/Filesystem/Security/AccessControlEntry.cs:40]: To disable for now. you will be abled to edit after merging phase 2.

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs (hunk 1 above)
Comment (0x5bfa) [reply src/Files.App/Filesystem/Security/AccessControlEntry.cs:49]: Don't think so. If we put AccessControlType.Deny directly, VS suggests you should add `_` switch case.

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs (hunk 2 above)
Comment (0x5bfa) [reply src/Files.App/Filesystem/Security/AccessControlEntry.cs:59]: As above.

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs (hunk 3 above)
Comment (0x5bfa) [reply src/Files.App/Filesystem/Security/AccessControlEntry.cs:357]: Disabled for now.

File: src/Files.App/Helpers/FileSecurityHelpers.cs (hunk 4 above)
Comment (0x5bfa) [reply src/Files.App/Helpers/FileSecurityHelpers.cs:21]: Yeah sorry. I will add some more info.

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (QuaintMako) [reply src/Files.App/Filesystem/Security/AccessControlEntry.cs:40]: May add a comment to keep track of it then

File: src/Files.App/Helpers/FileSecurityHelpers.cs (hunk 6 above)
Comment (0x5bfa) [reply src/Files.App/Helpers/FileSecurityHelpers.cs:182]: This is convention when calling Win32API.

File: src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs (hunk 7 above)
Comment (0x5bfa) [reply src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs:46]: Disabled for now.

File: src/Files.App/ViewModels/Properties/SecurityViewModel.cs (hunk 9 above)
Comment (0x5bfa) [reply src/Files.App/ViewModels/Properties/SecurityViewModel.cs:101]: Done

File: src/Files.App/Views/Properties/SecurityAdvancedPage.xaml (hunk 10 above)
Comment (0x5bfa) [reply src/Files.App/Views/Properties/SecurityAdvancedPage.xaml:695]: Disabled for now.

File: src/Files.App/Helpers/FileSecurityHelpers.cs (hunk 5 above)
Comment (0x5bfa) [reply src/Files.App/Helpers/FileSecurityHelpers.cs:44]: Done

File: src/Files.App/Helpers/FileSecurityHelpers.cs (hunk 4 above)
Comment (0x5bfa) [reply src/Files.App/Helpers/FileSecurityHelpers.cs:21]: Done

File: src/Files.App/Filesystem/Security/AccessControlEntry.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App/Filesystem/Security/AccessControlEntry.cs:40]: Sure then

File: src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs (hunk 8 above)
Comment (QuaintMako) [reply src/Files.App/ViewModels/Properties/SecurityAdvancedViewModel.cs:164]: Alright, let's keep it as is for now, I'll see if I can do a refactor for that!

**Review by QuaintMako — DISMISSED** [review]

File: src/Files.App/Strings/en-US/Resources.resw [hunk 12]
```diff
@@ -3208,4 +3208,16 @@
...
     <value>Location:</value>
   </data>
+  <data name="SecurityClickAdvancedPermissions" xml:space="preserve">
+    <value>Click 'Advanced permissions' to continue.</value>
+  </data>
+  <data name="SecurityRequireReadPermissions" xml:space="preserve">
+    <value>You must have Read permissions to view the properties of this object.</value>
+  </data>
+  <data name="SecuritySuggestToTakeOwnership" xml:space="preserve">
+    <value>To try taking ownership of the object, which includes permission to view its properties, click Change above.</value>
```
Comment (yaira2) [inline src/Files.App/Strings/en-US/Resources.resw:3218]:
```suggestion
    <value>To try taking ownership of the item, which includes permission to view its properties, click Change above.</value>
```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 13]
```diff
@@ -3208,4 +3208,16 @@
   <data name="Location" xml:space="preserve">
     <value>Location:</value>
   </data>
+  <data name="SecurityClickAdvancedPermissions" xml:space="preserve">
+    <value>Click 'Advanced permissions' to continue.</value>
+  </data>
+  <data name="SecurityRequireReadPermissions" xml:space="preserve">
+    <value>You must have Read permissions to view the properties of this object.</value>
```
Comment (yaira2) [inline src/Files.App/Strings/en-US/Resources.resw:3215]:
```suggestion
    <value>You must have Read permissions to view the properties of this item.</value>
```

File: src/Files.App/Strings/en-US/Resources.resw [hunk 14]
```diff
@@ -3208,4 +3208,16 @@
...
+  </data>
+  <data name="SecuritySuggestToTakeOwnership" xml:space="preserve">
+    <value>To try taking ownership of the object, which includes permission to view its properties, click Change above.</value>
+  </data>
+  <data name="SecurityUnableToDisplayPermissions" xml:space="preserve">
+    <value>Unable to display permissions for one or more errors</value>
```
Comment (yaira2) [inline src/Files.App/Strings/en-US/Resources.resw:3221]: This is a little confusing, I think we should rewrite it.

File: src/Files.App/Files.App.csproj [hunk 15]
```diff
@@ -82,6 +82,7 @@
         <Content Remove="Assets\FilesHome.png" />
     </ItemGroup>
     <ItemGroup>
+      <None Remove="Filesystem\Security\Principal.cs~RF5d03d9f.TMP" />
```
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App/Files.App.csproj:85]: Removed.

File: src/Files.App/Strings/en-US/Resources.resw (hunk 14 above)
Comment (0x5bfa) [reply src/Files.App/Strings/en-US/Resources.resw:3221]: I decided to display only 'Unable to display permissions' and Win32 formatted error in the next line.

image

**Review by yair100 — DISMISSED** [review]
LGTM

**Pull request merged**